Engine accepts paths to custom ruleset

ABaldwinHunter opened the pull request:
When PHPMD gets run, it accepts rule names, e.g.
codesize,unusedcode,naming

as a comma separated string.

It also accepts paths to a custome rule defined by the user mixed
in. For example:

```
  phpmd:
    enabled: true
    config:
      rulesets: "unusedcode,codesize,cleancode,design,naming,phpmd.xml"
```

where `phpmd.xml` contains a custom rule set up by user.

Previously, our engine failed to honor that config setup because it didn't
prefix `/code/` to the path. The result was 0 issues found across board.

This change fixes that behavior.

cc @codeclimate/review @leftees

Commit:
Engine accepts paths to custom ruleset

When PHPMD gets run, it accepts rule names, e.g.
codesize,unusedcode,naming

as a comma separated string.

It also accepts paths to a custome rule defined by the user mixed
in. For example:

```
  phpmd:
    enabled: true
    config:
      rulesets: "unusedcode,codesize,cleancode,design,naming,phpmd.xml"
```
where `phpmd.xml` contains a custom rule set up by user.

Previously, our engine failed to honor that config setup because it didn't
prefix `/code/` to the path. The result was 0 issues found across board.

This change fixes that behavior.

wfleming commented on Runner.php:
```diff
@@ -63,6 +65,20 @@ public function queuePaths($dir, $prefix = '', $exclusions = []) {
+        foreach ($configRulesets as &$r) {
+            if (!in_array($r, $officialPhpRulesets)) {
+             $r  = "/code/$r";
+             continue;
```
Is this actually needed? There's nothing else in the loop after this, so I don't think it accomplishes anything.

ABaldwinHunter replied:
I think you're right, I don't think it's needed. I can remove

wfleming commented on Runner.php:
```diff
@@ -78,12 +94,14 @@ public function run($files)
+            $rulesets = $this->prefixCodeDirectory($rulesets);
         }
 
+
```
:lipstick: extra line?

ABaldwinHunter commented:
Thanks @wfleming. Updated. Ready for another look!

wfleming commented on Runner.php:
```diff
@@ -63,6 +65,19 @@ public function queuePaths($dir, $prefix = '', $exclusions = []) {
+
+        foreach ($configRulesets as &$r) {
+            if (!in_array($r, $officialPhpRulesets)) {
+             $r  = "/code/$r";
```
:lipstick: looks like this needs to be indented another column?

wfleming commented on Runner.php:
```diff
@@ -78,10 +93,11 @@ public function run($files)
 
         if (isset($this->config['config']['rulesets'])) {
             $rulesets = $this->config['config']['rulesets'];
+            $rulesets = $this->prefixCodeDirectory($rulesets);
```
Any reason not to replace these two lines with `$rulesets = $this->prefixCodeDirectory($this->config['config']['rulesets'])`? Assigning to a var & then immediately reassigning it with a mutated version of itself is kind of smelly.

ABaldwinHunter replied:
okay

wfleming commented:
Two more suggestions (one of which PHP code sniffer also caught, so that's working, which is nice.). Other than that, LGTM.

Commit:
Fix style issues

ABaldwinHunter commented:
Thanks @wfleming.

I made a couple of minor style fixes in that file as well.

gdiggs commented on Runner.php:
```diff
@@ -63,6 +65,20 @@ public function queuePaths($dir, $prefix = '', $exclusions = []) {
+
+        foreach ($configRulesets as &$r) {
+            if (!in_array($r, $officialPhpRulesets)) {
+                $r  = "/code/$r";
```
I think it might be worth it to detect the case where someone enters an absolute path (ie. don't prepend `/code` if the string starts with `/`)

ABaldwinHunter replied:
:+1:

Commit:
Don't prefix /code/ for absolute paths

If a user provides an absolute path to ruleset file, we
won't prepend `/code/`. This behavior seems intuitive and also
prevents breakage with current codeclimate-php configs which may have
come to use `/code` as a workaround.

ABaldwinHunter commented:
@GordonDiggs @wfleming updated. Ready for a (hopefully) final look!

wfleming commented:
:+1:
